wsmaster/che-core-api-workspace/src/main/java/org/eclipse/che/api/workspace/server/WorkspaceManager.java

```diff
@@ -487,6 +487,29 @@ public void createSnapshot(String workspaceId) throws NotFoundException, ServerE
         return machineManager.getSnapshots(sessionUser().getId(), workspaceId);
     }
 
+    /**
+     * Adds machine into workspace runtime.
+     *
+     * @param machine
+     *         machine to add to specified runtime
+     * @throws ConflictException
+     *         if given machine already exists in specified runtime
+     *         or if machine is dev-machine
+     * @throws NotFoundException
+     *         if machine's workspace hasn't runtime
+     */
+    public void injectMachine(MachineImpl machine) throws ConflictException, NotFoundException, ServerException {
+        if (machine.getConfig().isDev()) {
+            throw new ConflictException("Cannot add another dev-machine");
```

Comment:
The message is not informative enough,  you should identify machine along with the workspace.